prusti-encoder/src/encoders/mir_builtin.rs

```diff
@@ -77,193 +76,277 @@ impl TaskEncoder for MirBuiltinEncoder {
+                    let e_l_ty = deps.require_ref::<crate::encoders::TypeEncoder>(
+                        l_ty,
+                    ).unwrap();
+                    let e_r_ty = deps.require_ref::<crate::encoders::TypeEncoder>(
+                        r_ty,
+                    ).unwrap();
+                    let e_rvalue_ty = deps.require_ref::<crate::encoders::TypeEncoder>(
+                        rvalue_ty,
                     ).unwrap();
+                    let prim_rvalue_ty = e_rvalue_ty.expect_prim();
 
-                    let name = vir::vir_format!(vcx, "mir_unop_neg_{}", int_name(*ty));
-                    let arity = UnknownArity::new(vcx.alloc_slice(&[ty_out.snapshot]));
+                    let name = vir::vir_format!(vcx, "mir_binop_{op:?}_{}_{}", int_name(l_ty), int_name(r_ty));
+                    let arity = UnknownArity::new(vcx.alloc_slice(&[e_l_ty.snapshot, e_r_ty.snapshot]));
                     let function = FunctionIdent::new(name, arity);
                     deps.emit_output_ref::<Self>(task_key.clone(), MirBuiltinEncoderOutputRef {
                         function,
                     });
-                    /* function mir_unop_neg(arg: s_I32): s_I32 {
-                        cons(-val(arg))
-                    } */
+                    let val = prim_rvalue_ty.prim_to_snap.apply(vcx,
+                        [vcx.alloc(vir::ExprData::BinOp(vcx.alloc(vir::BinOpData {
+                            kind: vir::BinOpKind::from(op),
+                            lhs: e_l_ty.expect_prim().snap_to_prim.apply(vcx,
+                                [vcx.mk_local_ex("arg1")],
+                            ),
+                            rhs: e_r_ty.expect_prim().snap_to_prim.apply(vcx,
+                                [vcx.mk_local_ex("arg2")],
+                            ),
+                        })))]
+                    );
+                    use mir::BinOp::*;
+                    let (pre, val) = match op {
+                        // Overflow well defined as wrapping
+                        Add | Sub | Mul | Shl | Shr =>
+                            (None, Self::get_wrapped_val(vcx, val, prim_rvalue_ty.prim_type)),
+                        // Undefined behavior to overflow (need precondition)
+                        AddUnchecked | SubUnchecked | MulUnchecked | ShlUnchecked | ShrUnchecked => {
+                            let wrapped_val = Self::get_wrapped_val(vcx, val, prim_rvalue_ty.prim_type);
+                            let pre = vcx.alloc(vir::ExprData::BinOp(vcx.alloc(vir::BinOpData {
+                                kind: vir::BinOpKind::CmpEq,
+                                lhs: val,
+                                rhs: wrapped_val,
+                            })));
+                            (Some(pre), val)
+                        }
+                        // Cannot overflow
+                        Div | Rem | BitXor | BitAnd | BitOr | Eq | Lt | Le | Ne | Ge | Gt | Offset =>
```

Comment:
What about division by `-1`?

Reply:
Hmm good point. It seems that Rust guarantees that this never happens, even in release ([playground](https://play.rust-lang.org/?version=stable&mode=release&edition=2021)). Not sure if we want to re-check this

Reply:
i.e. this is the same as division by `0` where we rely on Rust inserting an `assert` terminator before the division rather than rechecking things, or for overflows where Rust adds and `assert` after